docs/design/datacontracts/Metadata.md

```diff
@@ -0,0 +1,286 @@
+    #region MethodTable inspection APIs
+    public virtual MethodTableHandle GetMethodTableHandle(TargetPointer targetPointer) => throw new NotImplementedException();
+
+    public virtual TargetPointer GetModule(MethodTableHandle methodTable) => throw new NotImplementedException();
+    // A canonical method table is either the MethodTable itself, or in the case of a generic instantiation, it is the
+    // MethodTable of the prototypical instance.
+    public virtual TargetPointer GetCanonicalMethodTable(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual TargetPointer GetParentMethodTable(MethodTableHandle methodTable) => throw new NotImplementedException();
+
+    public virtual uint GetBaseSize(MethodTableHandle methodTable) => throw new NotImplementedException();
+    // The component size is only available for strings and arrays.  It is the size of the element type of the array, or the size of an ECMA 335 character (2 bytes)
+    public virtual uint GetComponentSize(MethodTableHandle methodTable) => throw new NotImplementedException();
+
+    // True if the MethodTable is the sentinel value associated with unallocated space in the managed heap
+    public virtual bool IsFreeObjectMethodTable(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual bool IsString(MethodTableHandle methodTable) => throw new NotImplementedException();
+    // True if the MethodTable represents a type that contains managed references
+    public virtual bool ContainsPointers(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual bool IsDynamicStatics(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumMethods(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumInterfaces(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumVirtuals(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumVtableSlots(MethodTableHandle methodTable) => throw new NotImplementedException();
+
+    // Returns an ECMA-335 TypeDef table token for this type, or for its generic type definition if it is a generic instantiation
+    public virtual uint GetTypeDefToken(MethodTableHandle methodTable) => throw new NotImplementedException();
+    // Returns the ECMA 335 TypeDef table Flags value (a bitmask of TypeAttributes) for this type,
+    // or for its generic type definition if it is a generic instantiation
+    public virtual uint GetTypeDefTypeAttributes(MethodTableHandle methodTable) => throw new NotImplementedException();
+    #endregion MethodTable inspection APIs
+```
+
+## Version 1
+
+The `MethodTable` inspection APIs are implemented in terms of the following flags on the runtime `MethodTable` structure:
+
+``` csharp
+internal partial struct Metadata_1
+{
+    [Flags]
```

Comment:
> I'd prefer to avoid these types.

Avoid them how?  This is how CoreCLR stores the bits.  this is an `internal` part of Metadata version 1. It doesn't leak out into the public `IMetadata` contract surface

> > We should do a bit better here. Is this the data descriptors or th contract? I'm struggling a bit with terms here.

There are three terms, in my mind:

* data descriptor: describes the layout of structures for a particular (version of a) runtime.  A contract implementation (`Metadata_1`) has a dependency on some data descriptors (that is - a runtime that provides `Metadata` version 1 is expected to provide a descriptors named `MethodTable` and `EEClass`)
* contract backend: the interpretation of the data in the target memory for a particular version of a contract. This is the `private` and `internal` fields/methods/structs of a particular contract (`Metadata_1` in our case)
* contract frontend: this is the `public` API surface provided by a contract (for example the `IMetadata` interface) - this is how you interact with the contract from the outside, for example to implement the SOS DAC APIs. 

Examples:

1. If CoreCLR changes where the `dwFlags` and `dwFlags2` fields are inside `MethodTable`, that just requires a change of the data descriptor.  no version bumps are necessary. an updated data descriptor embedded in the runtime will always be in sync.
2. If CoreCLR moves some flags or fields from `MethodTable` to `EEClass` but conceptually they represent the same logic, that will require a contract version bump and some of the `internal` members of `Metadata_vNext` will need to change, but the public contract API will stay the same.
3. If .NET 17 adds a new kind of type (for example some kind of zero-sized types like in Rust, that only exist for type computation) and they're relevant for diagnostic APIs, we would add some new accessor/query methods to `IMetadata` (with some kind of default behavior for older runtimes)